keps/sig-node/20190806-memory-manager.md

```diff
@@ -0,0 +1,264 @@
+
+- [CPU Manager][cpu-manager] is a feature provides a solution for CPU Pinning based on cgroups cpuset subsystem, it also offer topology hint to Topology Manager.
+
+- [Device Manager][device-manager] is a one of features provides topology hint for Topology Manager. The main objectivity of Device Manager is allow vendors to adverise their resources to kubelet.
+
+- [Hugepages][hugepages] is a feature allows container to consume pre-allocated hugepages with pod isolation of huge pages.
+
+- [Node Allocatable Feature][node-allocatable-feature] is a feature helps to reserve compute resources to prevent resource starvation. The kube-reserved and system-reserved is used to reserve resources for kubelet and system(OS, etc). Now following resources are supported to reserve: cpu, memory, ephemral storage.
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- Guarantee alignment of resources for isolating container's Memory and Hugepages with CPU and I/O devices.
+
+- Provide topology hint for Memory and Hugepages for Topology manager.
+
+## Non-Goals
+- Updating scheduler, pod spec is out of scope at this point.
+
+- This proposal only focus on Linux based system.
+
+## User Stories
+
+### Story 1 : Networking Acceleration using DPDK
+
+- DPDK(Data Plane Development Kit) is set of libraries to accelerate packet processing on userspace. DPDK requires dedicated resources and alignment of resources under NUMA awareness. There should be a way to gurantee reserve Memory and Hugepages with other computing resources on desinated Socket for containerized VNF using DPDK.
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+A new component of Kubelet enables NUMA-awareness for memory and hugepages. The main roles of this component are listed below.
+
+
+#### Calculate NUMA node affinity for memory and hugepages
+>NUMA node affinity represents that which NUMA node has enough capacity of memory and/or hugepages for a container. To calculate affinity Memory Manager takes capacity of memory and pre-allocated hugepages per NUMA node except system and Kubenetes reserved capacity by Node Allocatable feature. Then when pod admit is requested, Memory Manager checks resources availablity of each NUMA nodes and reserves resources internally. 
+- Node affinity of memory can be calcuated by below formulas.
+  - Allocatable Memory of NUMA node = Total Memory of NUMA node - Hugepages - system reserved - kubernetes reserved.
```

Comment:
@bg-chun How can we guarantee that the system reserved memory will use the memory only from some specific subset of NUMA nodes in an equal manner? Without it, you can not guarantee correct calculation for the `Allocatable Memory of NUMA node`.

Reply:
`system reserved`, `kuberntes reserved` fields in the formula, these are per node(worker node) value, not a numa node value.
These are equal to `--system-reserved=memory=xxxmi` and `--kube-reserved=memory=xxxmi` fields from [node allocatable feature](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/).

My original intention was a calculating reliable topology hints through having loss of memory usage at a worker node 

For example, 1) we have two sockets server, 2) total memory is 100GB(50GB per socket) 3) server has 10GB of 1G hugepages on each sockets(total 20GB) 4) we set 4GB for `system-reserved` and 2GB for `kube-reserve`.

In this example, allocatable memory for numa node 0 will be 36GB.
36GB(`allocatable memory of numa node 0`) = 50GB(`total memory of socket`) - 10GB(`hugepages`) - 4GB(`total reserved memory for system-reserved`) - 2GB(`total reserved memory for kube-reserved`).

Then allocatable memory for numa node 1 will be also 36GB.
so eventually, we lost 6GB.
But we can calculate topology hint precisely eventhough we have loss of memory usage.

I hope we find a better way, but I haven't  got a better idea yet :(

Reply:
Honestly i wish someone give me a briliant idea to address this issue.

Reply:
Also, this issue leads me to consider proposing this feature as an experimental feature.

Reply:
If we will have also `--system-reserved` huge pages, we will waste a huge amount of memory, maybe at least for huge pages, we can provide additional kubelet argument that will give the possibility to specify from what NUMA node do you want to reserve huge pages and it will be admin responsibility to configure it correct

Reply:
> 
> 
> we can provide additional kubelet argument that will give the possibility to specify from what NUMA node do you want to reserve huge pages

That's what I meant on this [comment](https://github.com/kubernetes/enhancements/pull/1203/files#r355439395).
```
For hugepages, my idea is adding a new flag to reserve hugepages per numa node.
I think we can extend this approach(kubernetes/kubernetes#83541) later.
```
I think it is better to write about the way of `providing additional kubelet argument` on this KEP.
Do you have any idea?

BTW I didn't update this KEP for a while. 
I think it is time to update and fix some types.
I will go a vacation 12/13 ~ 12/18.
I will update this KEP before I go.

Reply:
> That's what I meant on this [comment](https://github.com/kubernetes/enhancements/pull/1203/files#r355439395).

Sorry, I missed it.

> I think it is better to write about the way of `providing additional kubelet argument` on this KEP.
> Do you have any idea?

I think it will be best to explicitly describe current limitations of memory and huge pages calculation, and also describe how we possibly can overdo it.


> I will update this KEP before I go.

Great, a lot of thanks and have a good vacation!

Reply:
I tried to update entire content, but I couldn't, fortunately I updated key parts of major changes.
I added the concept of safety zone and reserved zone to manage memory at the NUMA node level.
I will update these parts and appendix(DPDK and Linux OOM system) after vacation.